src/v/cluster/config_manager.cc

```diff
@@ -540,23 +549,16 @@ ss::future<> config_manager::reconcile_status() {
 
     try {
         if (failed || should_send_status()) {
-            // * we were dirty & failed to send our update, sleep until retry
-            // OR
-            // * our status updated while we were sending, wait a short time
-            //   before sending our next update to avoid spamming the leader
-            //   with too many set_status RPCs if we are behind on seeing
-            //   updates to the controller log.
-            co_await ss::sleep_abortable(status_retry, _as.local());
+            co_await _reconcile_wait.wait(status_retry);
         } else {
             // We are clean: sleep until signalled.
-            co_await _reconcile_wait.wait();
+            co_await _reconcile_wait.wait(
+              [this]() { return should_send_status(); });
```

Comment:
in order to get the prompt shutdown when `_as` is activated, do you also need to signal the condition variable on shutdown?

---

separately, i'm glad you caught the missed wake-up issue. i'd even wonder if the cv::wait with the predicate could benefit from a timeout variant just to avoid unfortunate bugs in the future that result in dead lock.

that's a big downside for condition variable. one pattern that avoids that is to use a semaphore initialized to 0.

```
semaphore cv{0};

wait:
  co_await cv.wait(timeout);
  cv.consume(cv.available_units());

signal:
  co_await cv.signal();
```

Reply:
> in order to get the prompt shutdown when _as is activated, do you also need to signal the condition variable on shutdown?

I think the current implementation of shutdown relies on _cv.broken() which makes wait() throw and `broken_cv` is swallowed and the next iteration is not scheduled because it is predicated on _as. That seems to work, no?

> separately, i'm glad you caught the missed wake-up issue. i'd even wonder if the cv::wait with the predicate could benefit from a timeout variant just to avoid unfortunate bugs in the future that result in dead lock.

Seastar has that variant. That'd be an application bug where we never signal the CV and a timeout probably masks the bug because we see periodic updates like a timed wait. In all other cases I think just having a predicate will do.

```
  template<typename Clock = typename timer<>::clock, typename Duration = typename Clock::duration, typename Pred>
    SEASTAR_CONCEPT( requires seastar::InvokeReturns<Pred, bool> )
    future<> wait(std::chrono::time_point<Clock, Duration> timeout, Pred&& pred) noexcept {
```